### DISCUSSION TITLE
Charge now remote service missing

### DISCUSSION THREAD

**Discussion by DBHAHA** [bimmerconnected/bimmer_connected#440] [Q&A] [answered]
Hi
If I read PR #395 correctly, charge_now was added as a remote service at the beginning of the year. However, the feature doesn't show up in Home Assistant, and it isn't listed as a button in "button.py" under custom components? Might I be doing something wrong, or isn't the feature fully implemented yet? 

I'd would love to use this feature, as it would make it possible to charge my 530e at the cheapest possible time using Home Assistant - instead of manually adjusting the charging times in the myBMW-app every day 😀

PS: Thanks for a brilliant integration
![2022-05-24 21_50_41-Settings – Home Assistant](https://user-images.githubusercontent.com/76906154/170120583-6c931389-3706-4aa4-992f-0af02772ba8d.jpg)
!

**Comment by rikroe** [comment, marked as answer]
You're right, it is missing at the moment.
It was included in the library, but not in the HA component.

I'll put it into our backlog, shouldn't be a big effort. 

BTW: Does the MyBMW app also provide a "Stop Charging" service? Don't have an EV so cannot check this with my app.

**Reply by DBHAHA** [reply]
Sounds great, thanks for the quick reply.

There unfortunately isn't a "Stop Charging" button in the app, but you can change charging modes. It alternates between "charge now" and "charge in time slot". Switching back to a time slot set to e.g. 00:00 AM - 00:01 AM would almost be the same as "stop charging". 

I assume that this is the underlying function of the feature here:

8.17.3 bimmer_connected.vehicle.charging_profile
Models the charging profiles of a vehicle.
class bimmer_connected.vehicle.charging_profile.ChargingMode
Charging mode of electric vehicle.
DELAYED_CHARGING = 'delayedCharging'
IMMEDIATE_CHARGING = 'immediateCharging' 

[Charging management features in the My BMW app - BMW How-To](https://youtu.be/NrTCr9oMvWA)

![2022-05-24 22_32_10-(704) Charging management features in the My BMW app - BMW How-To - YouTube](https://user-images.githubusercontent.com/76906154/170127016-9b2a9941-c5cf-4c6a-be10-5de44d1adc12.jpg)

**Reply by darthcalamus** [reply]
The app also allows for stop charging.
It also allows to set the power limit - especially usefull for PV charging

**Reply by martinarva** [reply]
Would be nice to also have the scheduling part to HA.

**Reply by RutgerBeyen** [reply]
waiting for this feature as well

**Comment by DBHAHA** [comment]
Hi @rikroe 

Did you have an opportunity to work on the backlog regarding the charge_now service?
Thank you the work so far.

**Comment by RutgerBeyen** [comment]
Hi @rikroe,

Just to clarify that this is not about the remote service 'charge_now'.
The request is about controlling the ChargingMode as modeled in https://github.com/bimmerconnected/bimmer_connected/blob/master/bimmer_connected/vehicle/charging_profile.py
The BMW app allows for switching between IMMEDIATE_CHARGING and DELAYED_CHARGING, whereby the latter is controlled by a user-defined CHARGING_WINDOW (to set for example your local reduced energy-price timeframe.)
I'd be happy to supply information or run tests using my PHEV.

Thanks!

**Reply by rikroe** [reply]
For any of that type of thing (i.e. adding more remote services like changing `CHARGING_WINDOW`) you will need to monitor & log the commands sent by the BMW App on your phone to BMW servers.

You'll need a separate, rooted/jailbroken Android phone or iPhone as well as a local PC or small server to run e.g. mitmproxy.

You have something like that available or are able to set it up?

**Reply by RutgerBeyen** [reply]
Hi @rikroe,
Yes I have a rooted android emulator and the BMW app running.
However as soon as I start capturing traffic, the BMW app refuses any connection. I might need some help here in disabling this verification. I'm not sure if I understand the 'Ghidra' and 'Frida' part of the reverse-engineering guide correct.
Thanks!

**Comment by RutgerBeyen** [comment]
Hi @rikroe,

Got it working this morning.
Post calls are being done to: https://cocoapi.bmwgroup.com/eadrax-crccs/v1/vehicles/VIN/charging-profile

Request body when switching to 'charge immediately':
```json
{
  "servicePack": "ATM",
  "chargingMode": {
    "type": "CHARGING_IMMEDIATELY",
    "startTimeSlot": "0001-01-01T22:05:00.000",
    "endTimeSlot": "0001-01-01T06:55:00.000",
    "timerChange": "NO_CHANGE",
    "chargingPreference": "NO_PRESELECTION"
  },
  "departureTimer": {
    "type": "TWO_DEPARTURE_TIMER",
    "weeklyTimers": [
      {
        "id": 1,
        "time": "0001-01-01T09:00:00.000",
        "daysOfTheWeek": [],
        "timerAction": "ACTIVATE"
      },
      {
        "id": 2,
        "time": "0001-01-01T15:00:00.000",
        "daysOfTheWeek": [],
        "timerAction": "DEACTIVATE"
      }
    ]
  },
  "isPreconditionForDepartureActive": false
}
```

Switching to 'charge in window':
```json
{
  "servicePack": "ATM",
  "chargingMode": {
    "type": "TIME_SLOT",
    "startTimeSlot": "0001-01-01T22:05:00.000",
    "endTimeSlot": "0001-01-01T18:55:00.000",
    "timerChange": "NO_CHANGE",
    "chargingPreference": "CHARGING_WINDOW"
  },
  "departureTimer": {
    "type": "TWO_DEPARTURE_TIMER",
    "weeklyTimers": [
      {
        "id": 1,
        "time": "0001-01-01T09:00:00.000",
        "daysOfTheWeek": [],
        "timerAction": "ACTIVATE"
      },
      {
        "id": 2,
        "time": "0001-01-01T15:00:00.000",
        "daysOfTheWeek": [],
        "timerAction": "DEACTIVATE"
      }
    ]
  },
  "isPreconditionForDepartureActive": false
}
```

Looks like it is the 'chargingMode' block which is modified.

The response is an eventId which can be queried for its status. (https://cocoapi.bmwgroup.com/eadrax-vrccs/v3/presentation/remote-commands/eventStatus?eventId=ID)

**Reply by rikroe** [reply]
Very nice!

I hope that `servicePack` and `departureTimer` can always be extracted from the values retrieved for the vehicle (if the BMW APIs are open again, that is).

If so, it should be fairly easy integrating this into the library. We already query the `eventId`, so that should work seamlessly.

After integration in the library, how would we include something like this in HA?
Two separate buttons? A `switch` entity?

**Comment by RutgerBeyen** [comment]
Checking with a fingerprint and `servicePack` along with the existing `departureTimer` is returned (saved in file VIN_bmw_0.json):
```json
{
  "chargeAndClimateTimerDetail": {
    "chargingMode": {
      "chargingPreference": "CHARGING_WINDOW",
      "endTimeSlot": "0001-01-01T06:55:00",
      "startTimeSlot": "0001-01-01T22:05:00",
      "type": "TIME_SLOT"
    },
    "departureTimer": {
      "type": "TWO_DEPARTURE_TIMER",
      "weeklyTimers": [
        {
          "daysOfTheWeek": [],
          "id": 1,
          "time": "0001-01-01T09:00:00",
          "timerAction": "ACTIVATE"
        },
        {
          "daysOfTheWeek": [],
          "id": 2,
          "time": "0001-01-01T15:00:00",
          "timerAction": "DEACTIVATE"
        }
      ]
    },
    "isPreconditionForDepartureActive": false
  },
  "servicePack": "ATM"
}
```
A `switch`, two seperate buttons or a `Select` would do just fine I guess. Whatever is the easiest to implement

**Comment by rikroe** [comment]
Could you please provide a full fingerprint? There is no occurrence of `chargeAndClimateTimerDetail` in our [test responses](https://github.com/bimmerconnected/bimmer_connected/tree/master/test/responses) so the API seems to have changed.

**Comment by RutgerBeyen** [comment]
Sure!
[fingerprint.zip](https://github.com/bimmerconnected/bimmer_connected/files/9615730/fingerprint.zip)

`# bimmerconnected fingerprint **** **** rest_of_world
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:bimmer_connected.account:Getting vehicle list
DEBUG:bimmer_connected.api.authentication:Authenticating with MyBMW flow for North America & Rest of World.
DEBUG:httpx._client:HTTP Request: GET https://cocoapi.bmwgroup.com/eadrax-ucs/v1/presentation/oauth/config "HTTP/1.1 200 OK"
DEBUG:httpx._client:HTTP Request: POST https://customer.bmwgroup.com/gcdm/oauth/authenticate "HTTP/1.1 200 OK"
DEBUG:httpx._client:HTTP Request: POST https://customer.bmwgroup.com/gcdm/oauth/authenticate "HTTP/1.1 302 Found"
DEBUG:httpx._client:HTTP Request: POST https://customer.bmwgroup.com/gcdm/oauth/token "HTTP/1.1 200 OK"
DEBUG:httpx._client:HTTP Request: GET https://cocoapi.bmwgroup.com/eadrax-vcs/v2/vehicles "HTTP/1.1 200 OK"
DEBUG:httpx._client:HTTP Request: GET https://cocoapi.bmwgroup.com/eadrax-vcs/v2/vehicles "HTTP/1.1 200 OK"
DEBUG:httpx._client:HTTP Request: GET https://cocoapi.bmwgroup.com/eadrax-vcs/v2/vehicles/****/state "HTTP/1.1 200 OK
WARNING:bimmer_connected.models:'TWO_TIMES_TIMER' is not a valid 'TimerTypes'
DEBUG:httpx._client:HTTP Request: GET https://cocoapi.bmwgroup.com/eadrax-crccs/v1/vehicles/****?fields=charging-profe&has_charging_settings_capabilities=true "HTTP/1.1 200 OK"
fingerprint of the vehicles written to /vehicle_fingerprint/2022-09-21_12-28-24`

**Reply by rikroe** [reply]
Oh, this is interesting!

The data needed seems to come from `eadrax-crccs/v1/vehicles/VIN?fields=charging-profe&has_charging_settings_capabilities=true`, which we don't parse in the library yet.

We just know it is there but didn't even find out when it will work on a car and when it doesn't. I'll have to dig deeper into this.

**Reply by rikroe** [reply]
Ok, came up with a scrappy solution - not sure if that's what we want (also the question is if more interesting data can be retrieved from the above API).

Please try the following:
Install my test branch using `pip install git+https://github.com/rikroe/bimmer_connected.git@charging-preference` (see changes here https://github.com/bimmerconnected/bimmer_connected/compare/master...rikroe:bimmer_connected:charging-preference).

Then, execute one of the remote services using the following (or however you would like to execute them):
```py
import asyncio

from bimmer_connected.account import MyBMWAccount
from bimmer_connected.api.regions import get_region_from_name
from bimmer_connected.vehicle.charging_profile import ChargingPreferences

USERNAME = ""
PASSWORD = ""
REGION = get_region_from_name("rest_of_world")
VIN = ""

myBMW = MyBMWAccount(USERNAME, PASSWORD, REGION)
asyncio.run(myBMW.get_vehicles())


asyncio.run(myBMW.get_vehicle(VIN).remote_services.trigger_charging_mode_change(ChargingPreferences.CHARGING_WINDOW))
# or
# asyncio.run(myBMW.get_vehicle(VIN).remote_services.trigger_charging_mode_change(ChargingPreferences.NO_PRESELECTION))
```

**Reply by RutgerBeyen** [reply]
Actually, after some small fixes, it works pretty well!
The URL for getting the current charging-profile and updating it is slightly different.
I created a PR with fixes to the code of the test branch: https://github.com/rikroe/bimmer_connected/pull/2

**Reply by rikroe** [reply]
Ah very cool, thanks!

I'll need some time thinking about how to fully integrate this in the library - yes, it works but having the GET endpoint only called when actually changing something feels out of place compared to the rest of the lib.

I feel it should rather be included in the actual `get_vehicles()` code and update `MyBMWVehicle.ChargingProfile` - however that usually gets its data from the default `state` endpoint and some/most(?) data seems to be duplicated?

So the question would be is any of the additional data from the new endpoint needed/interesting or we just need it to update?

Also, how does this compare with the actual `CHARGE_NOW` remote service? Feels somehow duplicated...

**Reply by RutgerBeyen** [reply]
ideally all necessary data is retrieved at the polling interval and the correct body is reassembled for updating the charging profile.
It probably needs another value for the `["chargingMode"]["timerChange"]` field when you update the timers, I still need to figure that out.
(Some cars have the TWO_TIMES_TIMER, others have a TWO_WEEKS_TIMER, might need different body payloads as well.)

The existing `CHARGE_NOW` remote service is not working on my PHEV. I get a 404 from the BMW servers. Perhaps that is a remote service specific for a BEV?

**Comment by fredriklj** [comment]
Interested in hearing if this effort is still ongoing? It looks like you have gotten quite far! Also seem to be on the wish list for quite a few.

I can definitely help with some testing on an i4 if required. One feature I'd be particularly interested in is to be able to control targetSoc. It's recommended setting is 80% but at some times you may want to charge to 100% and other times perhaps lower to get just enough juice.

I haven't got access to a rooted phone, so unfortunately I do not know the exact call for this operation, but it is in the fingerprint files that @RutgerBeyen posted.

**Reply by rikroe** [reply]
@fredriklj there is now a way where you do not need a rooted Android anymore - it should also work with non-rooted devices. Is this something you have access to?
If yes I can see if I am able to provide you with instructions how to get this running so you can get some actual requests for us.

**Reply by fredriklj** [reply]
I have an old Samsung Galaxy J7 from 2018 which I can play with, not sure however if it is up for running the latest app? It looks though like Rutger got it working in an emulator, I thought I'd ask him how he did.

But, please, provide me the instructions and I'll see what I can do! You'll find my e-mail address on my profile.

**Comment by RutgerBeyen** [comment]
Since the myBMW app v2.11 the URLs have changed slightly:

Gettting the existing charging settings is now done via `https://cocoapi.bmwgroup.com/eadrax-crccs/v2/vehicles?fields=charging-profile&has_charging_settings_capabilities=True`
with an extra Header: `bmw-vin: VIN`

The response from this call can be POSTed to `https://cocoapi.bmwgroup.com/eadrax-crccs/v1/vehicles/VIN/charging-profile` for actually modifying the settings.
TargetSOC is not a parameter which I can control for my PHEV. Not in the car and not via the app.

**Reply by fredriklj** [reply]
Thanks for checking this. I'm not sure exactly sure what you get/post from the endpoint, but do you see there is a possibility to make an educated guess, considering the output from the fingerprinting? In your vin_bmw_0.json there is this dictionary:

`
  "chargingSettingsDetail": {
    "acLimit": {
      "current": {
        "unit": "A",
        "value": 0
      },
      "isUnlimited": false,
      "max": 32,
      "min": 0,
      "values": [
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        20,
        32
      ]
    },
    "chargingTarget": 100,
    "isUnlockCableActive": false,
    "minChargingTargetToWarning": 0
  },
 ...
`    

chargingTarget would be the one I'm after I think.

**Comment by kriscremers** [comment]
Following! Would like to give this a whirl testing wise if any further assistance is needed. Located in the EU, driving an iX with a load-balancing charging station that's dumber than the cobblestones it's standing next to. So some automation is definitely needed, especially with the new electricity regulations here.

**Comment by rikroe** [comment]
Just released a beta that contains most of the desired functionality: https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.13.0b1
Please help me out, test it and give me feedback in #520.
